
<launch>

  <!-- ROBOT MAPPING VERSION: use this with ROS bag demo_mapping.bag -->
  <!-- WARNING : Database is automatically deleted on each startup -->
  <!--           See "delete_db_on_start" option below... -->
    <arg name="rtabmap_args" default="--delete_db_on_start"/>
    <arg name="mapping_mode" default="true"/>

  <!-- Choose visualization -->
  <arg name="rviz" default="false" />
  <arg name="rtabmapviz" default="true" /> 
  
  <param name="use_sim_time" type="bool" value="True"/>
  
  <!-- Localization-only mode -->
  <arg name="localization"      default="false"/>
  <arg     if="$(arg localization)" name="rtabmap_args"  default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args"  default="--delete_db_on_start"/>

  <group ns="rtabmap">
    <!-- SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
	  <param name="frame_id"           type="string" value="base_footprint"/>
	  <param name="wait_for_transform" type="bool" value="true"/>
	
	  <param name="subscribe_depth" type="bool" value="true"/>
	  <param name="subscribe_scan"  type="bool" value="false"/>
	
	  <remap from="odom" to="/az3/base_controller/odom"/>
	  <remap from="scan" to="/jn0/base_scan"/>

	  <remap from="rgb/image"       to="/data_throttled_image"/>
  	  <remap from="depth/image"     to="/data_throttled_image_depth"/>
  	  <remap from="rgb/camera_info" to="/data_throttled_camera_info"/>
  	
	  <param name="rgb/image_transport"   type="string" value="compressed"/>
	  <param name="depth/image_transport" type="string" value="compressedDepth"/>
	
	  <!-- RTAB-Map's parameters: do "rosrun rtabmap rtabmap (double-dash)params" to see the list of available parameters. -->
	  <param name="RGBD/NeighborLinkRefining" type="string" value="true"/>  <!-- Do odometry correction with consecutive laser scans -->
	  <param name="RGBD/ProximityBySpace"     type="string" value="true"/>  <!-- Local loop closure detection (using estimated position) with locations in WM -->
	  <param name="RGBD/ProximityByTime"      type="string" value="false"/> <!-- Local loop closure detection with locations in STM -->
	  <param name="Reg/Strategy"              type="string" value="1"/>     <!-- 0=Visual, 1=ICP, 2=Visual+ICP -->
	  <param name="Vis/InlierDistance"        type="string" value="0.1"/>   <!-- 3D visual words correspondence distance -->
	  <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/> <!-- Optimize graph from initial node so /map -> /odom transform will be generated -->
	  <param name="Reg/Force3DoF"             type="string" value="true"/>
	  <param name="Grid/FromDepth"            type="string" value="false"/>
	  
	  <!-- localization mode -->
	  <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
	  <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
	  <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/>
   </node>
    
    <!-- Visualisation RTAB-Map -->
    <node if="$(arg rtabmapviz)" pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
  	  <param name="subscribe_depth"    type="bool" value="true"/>
      <param name="subscribe_scan"     type="bool" value="true"/>
      <param name="frame_id"           type="string" value="base_footprint"/>
      <param name="wait_for_transform" type="bool" value="true"/>
    
      <remap from="rgb/image"       to="/zed/rgb/image_rect_color"/>
      <remap from="depth/image"     to="/zed/depth/depth_registered"/>
      <remap from="rgb/camera_info" to="/zed/rgb/camera_info"/>
      <remap from="odom"            to="/odometry/filtered/sync"/>
      <remap from="scan"            to="/scan/sync"/>
      
      <param name="rgb/image_transport"   type="string" value="compressed"/>
      <param name="depth/image_transport" type="string" value="compressedDepth"/>
    </node>
  </group>
  
    <node pkg="nodelet" type="nodelet" name="zed" ns="zed" args="load zed_wrapper/ZEDWrapperNodelet /mapping_nodelet_manager">
        <param name="publish_tf"            value="false" />

        <param name="odometry_frame"        value="odom" />
        <param name="base_frame"            value="zed_center" />
        <param name="camera_frame"          value="zed_left_camera" />
        <param name="depth_frame"           value="zed_depth_camera" />

        <param name="svo_filepath"          value="" />

        <param name="zed_id"                value="0" />

        <param name="resolution"            value="3" />
        <param name="quality"               value="1" />
        <param name="sensing_mode"          value="0" />
        <param name="frame_rate"            value="60" />
        <param name="odometry_db"           value="" />
        <param name="openni_depth_mode"     value="0" />
        <param name="gpu_id"                value="-1" />
        <param name="confidence"            value="100" />
        <param name="depth_stabilization"   value="1" />

        <param name="rgb_topic"             value="rgb/image_rect_color" />
        <param name="rgb_raw_topic"         value="rgb/image_raw_color" />
        <param name="rgb_cam_info_topic"    value="rgb/camera_info" />

        <param name="left_topic"            value="left/image_rect_color" />
        <param name="left_raw_topic"        value="left/image_raw_color" />
        <param name="left_cam_info_topic"   value="left/camera_info" />

        <param name="right_topic"           value="right/image_rect_color" />
        <param name="right_raw_topic"       value="right/image_raw_color" />
        <param name="right_cam_info_topic"  value="right/camera_info" />

        <param name="depth_topic"           value="depth/depth_registered" />
        <param name="depth_cam_info_topic"  value="depth/camera_info" />

        <param name="point_cloud_topic"     value="point_cloud/cloud_registered" />

        <param name="odometry_topic"        value="odom" />
    </node>

    <param name="zed_description" textfile="$(find zed_wrapper)/urdf/zed.urdf" />
    <node name="zed_state_publisher" pkg="robot_state_publisher" type="state_publisher">
        <remap from="robot_description" to="zed_description" />
    </node>

    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_localization" clear_params="true">
        <param name="frequency" value="60"/>
        <param name="sensor_timeout" value="0.03"/>
        <param name="two_d_mode" value="true"/>
        <param name="publish_tf" value="true"/>

        <param name="odom0" value="/zed/odom"/>
        <rosparam param="odom0_config">[true, true, true,
                                        true, true, true,
                                        false,  false, false,
                                        false, false, false,
                                        false, false, false]</rosparam>

        <param name="imu0" value="/buggy/imu"/>
        <rosparam param="imu0_config">[false, false, false,
                                       false, false, false,
                                       false, false, false,
                                       true, true, true,
                                       true, true, true]</rosparam>
        <param name="imu0_remove_gravitational_acceleration" value="true"/>
        <param name="imu0_threshold" value="100"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="base_link_camera_link" args="0 0 0.35 0 0 0 base_link camera_link 20"/>
    <node pkg="tf" type="static_transform_publisher" name="camera_link_zed_center" args="0 0 0 0 0 0 camera_link zed_center 20"/>
    <node pkg="tf" type="static_transform_publisher" name="base_link_imu" args="0 0 0 0 0 3.14 base_link imu 20"/>


  <!-- Visualisation RVIZ -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_ros)/launch/config/demo_robot_mapping.rviz" output="screen"/>
   <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="standalone rtabmap_ros/point_cloud_xyzrgb">
    <remap from="rgb/image"       to="/data_throttled_image"/>
    <remap from="depth/image"     to="/data_throttled_image_depth"/>
    <remap from="rgb/camera_info" to="/data_throttled_camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <param name="rgb/image_transport"   type="string" value="compressed"/>
    <param name="depth/image_transport" type="string" value="compressedDepth"/>
    
    <param name="queue_size" type="int" value="10"/>
    <param name="voxel_size" type="double" value="0.01"/>
  </node>

</launch>
